### PR TITLE
New: multi-line option for the curly rule (fixes #1812)

### LINKED ISSUE #1812: curly: option to allow omission when there is only one statement *and* it is on the same line
Rationale: retains the advantages of the `curly` rule, while allowing compactness when the condition and its statement are trivial and not important enough to take up 3 lines (which may amount to 75% of the code of the function it's in).

Valid
```
if (foo) doSomething();
else doSomethingElse();

if (foo) {
  doSomething();
} else {
  doSomethingElse();
}
```

Invalid
```
if (foo)
  doSomething();
else
  doSomethingElse();

if (foo
  && bar) baz();

if (foo) foo(
  bar,
  baz);
```

Does it make sense? I'm willing to take a shot at coding it.

<bountysource-plugin>

---
Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/8424161-curly-option-to-allow-omission-when-there-is-only-one-statement-and-it-is-on-the-same-line?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by hgwood**
For the rationale and behavior of this new option, see #1812.

### CODE REVIEW

File: lib/rules/curly.js [hunk 1]
```diff
@@ -10,13 +10,24 @@
 
 module.exports = function(context) {
 
-    var multiOnly = (context.options[0] === "multi");
+    var multiOnly = (context.options.indexOf("multi") >= 0);
+    var allowOneLiners = (context.options.indexOf("allow-one-liners") >= 0);
```
Comment (nzakas) [inline lib/rules/curly.js]: We typically don't have multiple string options defined in this way. It either needs to replace "multi" or needs to be in an object with properties, such as:

```
curly: [2, "multi", {
    allowSingleLine: true
}
```

**Comment by vp-hwood** [comment]
Working on the documentation.

**Comment by hgwood** [comment]
Working on the documentation

**Comment by nzakas** [comment]
Great, thanks!

**Comment by hgwood** [comment]
I have noticed the current implementation wrongly warns about
```js
if (foo) bar();
else baz();
```
I'm working on that too.

**Comment by hgwood** [comment]
f0c1d4249fb7ae1fa07dab19d751cc5c4f8c45c0: Docs and rule updated, but branch coverage decreased a little bit. I'm looking into it.

**Comment by hgwood** [comment]
ea8848afcdf9baf9f7766bb42b6ca0223080ebe9: branch coverage on curly is back to 100%. Ready for comments or merge.

File: docs/rules/curly.md [hunk 2]
```diff
@@ -92,6 +92,52 @@ while (true) {
 }
 ```
 
+Alternatively, you can relax the rule to allow brace-less single-line `if`, ``else if`, `else`, `for`, `while`, or `do`, while still enforcing the use of curly braces for other instances. To do so, configure the rule as:
```
Comment (nzakas) [inline docs/rules/curly.md]: Double back tick is causing formatting issues.

**Comment by nzakas** [comment]
Looks good. Can you sign our CLA? http://eslint.org/cla this just gives us permission to use your code

**Comment by hgwood** [comment]
Done.

**Comment by nzakas** [comment]
Thanks!

**Pull request merged**